src/Bicep.Core/BicepCompiler.cs

```diff
@@ -116,13 +113,12 @@ private Compilation Create(SourceFileGrouping sourceFileGrouping)
 
         static IEnumerable<(BicepSourceFile, IDiagnostic)> GetDiagnosticsForModulesToRestore(SourceFileGrouping grouping, ImmutableHashSet<ArtifactResolutionInfo> originalArtifactsToRestore)
         {
-            var originalModulesToRestore = originalArtifactsToRestore.OfType<ArtifactResolutionInfo>();
-            foreach (var (module, sourceFile) in originalModulesToRestore)
+            foreach (var artifact in originalArtifactsToRestore)
             {
-                if (sourceFile is BicepSourceFile bicepFile &&
-                    DiagnosticForModule(grouping, module) is { } diagnostic)
+                if (artifact.Syntax is {} &&
```

Comment:
there are a couple of places where we make this check to decide if an artifact is implicit. I think it may be worth it to implement a `public bool IsImplicit()` instead

Reply:
This pattern is necessary for type analysis (the compiler needs to know the syntax is non-null). There's probably a clever way to accomplish that, but I didn't give it too much thought.